internal/provider/waypoint/data_source_waypoint_action_config.go

```diff
@@ -0,0 +1,229 @@
+					"custom": schema.SingleNestedAttribute{
+						Description: "Custom mode allows users to define the HTTP method, the request body, etc.",
+						Computed:    true,
+						Attributes: map[string]schema.Attribute{
+							"method": schema.StringAttribute{
+								Description: "The HTTP method to use for the request.",
+								Computed:    true,
+							},
+							"headers": schema.MapAttribute{
+								Description: "Key value headers to send with the request.",
+								Computed:    true,
+								ElementType: types.StringType,
+							},
+							"url": schema.StringAttribute{
+								Description: "The full URL this request should make when invoked.",
+								Computed:    true,
+							},
+							"body": schema.StringAttribute{
+								Description: "The body to be submitted with the request.",
+								Computed:    true,
+							},
+						},
+					},
+					/*"github": schema.ListNestedAttribute{
+						Description: "GitHub mode is configured to do various operations on GitHub Repositories.",
+						Optional:    true,
+					},
+					"agent": schema.ListNestedAttribute{
+						Optional: true,
+					},*/
+				},
+			},
+		},
+	}
+}
+
+func (d *DataSourceActionConfig) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	d.client = client
+}
+
+func (d *DataSourceActionConfig) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data ActionConfigResourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	client := d.client
+	if d.client == nil {
+		resp.Diagnostics.AddError(
+			"Unconfigured HCP Client",
+			"Expected configured HCP client. Please report this issue to the provider developers.",
+		)
+		return
+	}
+
+	projectID := client.Config.ProjectID
+	if !data.ProjectID.IsNull() {
+		projectID = data.ProjectID.ValueString()
+	}
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      projectID,
+	}
+
+	var actionCfg *waypoint_models.HashicorpCloudWaypointActionConfig
+	var err error
+
+	actionCfg, err = clients.GetActionConfig(ctx, client, loc, data.ID.ValueString(), data.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(err.Error(), "")
+		return
+	}
+
+	if actionCfg.ID != "" {
+		data.ID = types.StringValue(actionCfg.ID)
+	}
+	if actionCfg.Name != "" {
+		data.Name = types.StringValue(actionCfg.Name)
+	}
+	if actionCfg.Description != "" {
+		data.Description = types.StringValue(actionCfg.Description)
+	}
+	if actionCfg.ActionURL != "" {
+		data.ActionURL = types.StringValue(actionCfg.ActionURL)
+	}
```

Comment:
The problem that I've encountered with this is that in some cases we do not set these, so the provider has them in state as `null` but the API sends back the empty string `""`. In this case, we really do want to ignore the empty string because if we don't, we get an error like this: `When applying changes to hcp_waypoint_action_config.test, provider
        "provider[\"registry.terraform.io/hashicorp/hcp\"]" produced an unexpected
        new value: .action_url: was null, but now cty.StringVal("")`